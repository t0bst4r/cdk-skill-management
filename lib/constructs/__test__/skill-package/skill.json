{
  "manifest": {
    "manifestVersion": "1.0",
    "publishingInformation": {
      "distributionCountries": [],
      "locales": {
        "de-DE": {
          "description": "My Personal Smart Home",
          "examplePhrases": [],
          "keywords": [
            "Smart Home",
            "Lights",
            "Smart Devices"
          ],
          "name": "Test-Smart-Home-Skill",
          "summary": "My Personal Smart Home"
        }
      },
      "testingInstructions": "Sample Testing Instructions."
    },
    "permissions": [
      {
        "name": "alexa::async_event:write"
      }
    ],
    "apis": {
      "smartHome": {
        "protocolVersion": "3",
        "endpoint": {
          "uri": "//override"
        }
      }
    }
  }
}