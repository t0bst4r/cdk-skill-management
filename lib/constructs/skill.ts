import {IResource, Resource} from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {CfnSkill} from 'aws-cdk-lib/alexa-ask';
import {Role, ServicePrincipal} from 'aws-cdk-lib/aws-iam';
import {Asset} from 'aws-cdk-lib/aws-s3-assets';
import {SkillAuthenticationProps} from './skill-authentication-props';
import {ISecret} from 'aws-cdk-lib/aws-secretsmanager';
import {IStringParameter} from 'aws-cdk-lib/aws-ssm';

/**
 * Enumeration for different Alexa Skill types.
 */
export enum SkillType {
  CUSTOM = 'CUSTOM',
  SMART_HOME = 'SMART_HOME',
}

/**
 * Interface representing an Alexa Skill.
 */
export interface ISkill extends IResource {
  /** The unique ID of the Alexa Skill. */
  readonly skillId: string;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
}

/**
 * Base class for the Alexa Skill construct.
 */
abstract class SkillBase extends Resource implements ISkill {
  /**
   * The unique ID of the Alexa Skill.
   */
  public abstract readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public abstract readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public abstract readonly skillType: SkillType;
}

/**
 * Type alias for Alexa Skill package overrides.
 */
export type SkillPackageOverrides = CfnSkill.OverridesProperty;

/**
 * Interface representing an Alexa Skill package.
 */
export interface SkillPackage {
  /** The asset associated with the Alexa Skill package. */
  readonly asset: Asset;
  /** Overrides for the Alexa Skill package. */
  readonly overrides?: SkillPackageOverrides;
}

/**
 * Properties for creating an Alexa Skill.
 */
export interface SkillProps extends SkillAuthenticationProps {
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The vendor ID of the Alexa Skill. */
  readonly vendorId: string;
  /** The package information for the Alexa Skill. */
  readonly skillPackage: SkillPackage;
}

/**
 * Properties for creating an Alexa Skill Instance from Attributes.
 */
export interface SkillAttributes {
  /**
   * The unique ID of the Alexa Skill.
   */
  readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  readonly skillType: SkillType;
}

/**
 * Alexa Skill construct for managing an Alexa Skill via CloudFormation.
 */
export class Skill extends SkillBase {
  /**
   * Factory method to create an Alexa Skill from its ID.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skillId - The ID of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromSkillId(scope: Construct, id: string, skillId: string): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = skillId;

      get skillType(): SkillType {
        throw new Error('Access to skillType is not supported when using fromSkillId');
      }

      get skillStage(): string {
        throw new Error('Access to skillStage is not supported when using fromSkillId');
      }
    })(scope, id);
  }

  /**
   * Factory method to create an Alexa Skill from its attributes.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param attributes - The attributes of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromAttributes(scope: Construct, id: string, attributes: SkillAttributes): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = attributes.skillId;
      readonly skillType: SkillType = attributes.skillType;
      readonly skillStage: string = attributes.skillStage;
    })(scope, id);
  }

  private readonly packageRole: Role;
  private readonly resource: CfnSkill;

  /**
   * The unique ID of the Alexa Skill.
   */
  public readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public readonly skillType: SkillType;

  /**
   * Creates an instance of the Alexa Skill construct.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Alexa Skill properties.
   */
  constructor(scope: Construct, id: string, props: SkillProps) {
    super(scope, id);

    const authPropsCount = [
      props.authenticationConfiguration,
      props.authenticationConfigurationParameter,
      props.authenticationConfigurationSecret,
    ].filter(it => !!it).length;

    if (authPropsCount !== 1) {
      throw new Error('Exactly one authentication configuration needs to be provided!');
    }

    this.packageRole = new Role(this, 'PackageRole', {
      assumedBy: new ServicePrincipal('alexa-appkit.amazon.com'),
    });

    const authentication: CfnSkill.AuthenticationConfigurationProperty =
      this.authenticationSecret(props.authenticationConfigurationSecret) ??
      this.authenticationParameter(props.authenticationConfigurationParameter) ??
      props.authenticationConfiguration!;

    this.resource = new CfnSkill(this, 'Default', {
      skillPackage: {
        s3Bucket: props.skillPackage.asset.s3BucketName,
        s3Key: props.skillPackage.asset.s3ObjectKey,
        s3BucketRole: this.packageRole.roleArn,
        overrides: {
          ...props.skillPackage.overrides,
          manifest: {
            ...props.skillPackage.overrides?.manifest,
            publishingInformation: {
              category: props.skillType,
              ...props.skillPackage.overrides?.manifest?.publishingInformation,
            },
          },
        },
      },
      vendorId: props.vendorId,
      authenticationConfiguration: authentication,
    });
    this.resource.node.addDependency(this.packageRole);
    this.resource.node.addDependency(props.skillPackage.asset);
    props.skillPackage.asset.grantRead(this.packageRole);

    this.skillId = this.resource.ref;
    this.skillStage = props.skillStage;
    this.skillType = props.skillType;
  }

  private authenticationSecret(secret: ISecret | undefined): CfnSkill.AuthenticationConfigurationProperty | undefined {
    if (!secret) {
      return;
    }
    return {
      clientId: secret.secretValueFromJson('clientId').unsafeUnwrap(),
      clientSecret: secret.secretValueFromJson('clientSecret').unsafeUnwrap(),
      refreshToken: secret.secretValueFromJson('refreshToken').unsafeUnwrap(),
    };
  }

  private authenticationParameter(
    parameter: IStringParameter | undefined
  ): CfnSkill.AuthenticationConfigurationProperty | undefined {
    if (!parameter) {
      return;
    }
    return JSON.parse(parameter.stringValue);
  }
}
