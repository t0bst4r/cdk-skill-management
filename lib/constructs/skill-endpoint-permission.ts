import {Construct, IDependable} from 'constructs';
import {CfnPermission, IFunction} from 'aws-cdk-lib/aws-lambda';
import {ISkill, SkillType} from './skill';
import {AwsCustomResource, AwsCustomResourcePolicy, AwsSdkCall, PhysicalResourceId} from 'aws-cdk-lib/custom-resources';
import {PolicyStatement} from 'aws-cdk-lib/aws-iam';

/**
 * Properties for configuring a Skill Endpoint Permission.
 */
export interface SkillEndpointProps {
  /** The AWS Lambda function handler to configure the permission for. */
  readonly handler: IFunction;
}

/**
 * Interface representing a Skill Endpoint Permission.
 */
export interface ISkillEndpointPermission {
  /**
   * Configures a Skill Endpoint Permission for a specific skill.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skill - The skill to configure the permission for.
   * @returns An IDependable object representing the configured permission.
   */
  configureSkillId(scope: Construct, id: string, skill: ISkill): IDependable;
}

/**
 * Class for configuring and managing Skill Endpoint Permissions.
 */
export class SkillEndpointPermission extends Construct implements ISkillEndpointPermission {
  private readonly principals: Record<SkillType, string> = {
    [SkillType.CUSTOM]: 'alexa-appkit.amazon.com',
    [SkillType.SMART_HOME]: 'alexa-connectedhome.amazon.com',
  };

  private readonly handler: IFunction;
  private readonly permission: CfnPermission;
  private readonly policy: AwsCustomResourcePolicy;

  /**
   * Creates an instance of the Skill Endpoint Permission.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Skill Endpoint properties.
   */
  constructor(scope: Construct, id: string, props: SkillEndpointProps) {
    super(scope, id);

    this.permission = new CfnPermission(this, 'InitialSkillPermission', {
      functionName: props.handler.functionArn,
      principal: this.principals[SkillType.CUSTOM],
      action: 'lambda:InvokeFunction',
    });

    this.policy = AwsCustomResourcePolicy.fromStatements([
      new PolicyStatement({
        actions: ['lambda:RemovePermission', 'lambda:AddPermission'],
        resources: [props.handler.functionArn],
      }),
    ]);

    this.handler = props.handler;
  }

  /**
   * Configures a Skill Endpoint Permission for a specific skill.
   * This replaces the initially created permission with a new permission checking for the SkillId.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skill - The skill to configure the permission for.
   * @returns An IDependable object representing the configured permission.
   */
  public configureSkillId(scope: Construct, id: string, skill: ISkill): IDependable {
    const parent = new Construct(scope, id);
    new AwsCustomResource(parent, 'RemovePermission', {
      policy: this.policy,
      onCreate: this.removePermissionCall(parent, skill),
      onDelete: this.addPermissionCall(parent),
      installLatestAwsSdk: false,
    });
    new AwsCustomResource(parent, 'AddPermission', {
      policy: this.policy,
      onCreate: this.addPermissionCall(parent, skill),
      onDelete: this.removePermissionCall(parent, skill),
      installLatestAwsSdk: false,
    });
    parent.node.addDependency(this.permission, this.handler);
    return parent;
  }

  private removePermissionCall(scope: Construct, skill: ISkill): AwsSdkCall {
    return {
      service: 'Lambda',
      action: 'removePermission',
      parameters: {
        FunctionName: this.handler.functionArn,
        StatementId: this.permission.attrId,
      },
      physicalResourceId: PhysicalResourceId.of(`RemovePermission-${scope.node.addr}-${skill.skillId}`),
    };
  }

  private addPermissionCall(scope: Construct, skill?: ISkill): AwsSdkCall {
    return {
      service: 'Lambda',
      action: 'addPermission',
      parameters: {
        FunctionName: this.handler.functionArn,
        StatementId: this.permission.attrId,
        Principal: this.principals[skill?.skillType ?? SkillType.CUSTOM],
        Action: 'lambda:InvokeFunction',
        EventSourceToken: skill?.skillId,
      },
      physicalResourceId: PhysicalResourceId.of(`AddPermission-${scope.node.addr}-${skill?.skillId ?? 'no-skill'}`),
    };
  }
}
