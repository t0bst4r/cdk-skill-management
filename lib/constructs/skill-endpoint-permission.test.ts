import {App, Stack} from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import {Skill, SkillType} from './skill';
import {Function} from 'aws-cdk-lib/aws-lambda';
import {SkillEndpointPermission} from './skill-endpoint-permission';

test('SkillEndpointPermission', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  const endpointFunction = Function.fromFunctionName(stack, 'EndpointFunction', 'MyEndpointFunction');

  const endpointPermission = new SkillEndpointPermission(stack, 'EndpointPermission', {
    handler: endpointFunction,
  });
  const skill = Skill.fromAttributes(stack, 'Skill', {
    skillId: 'My-Skill-Id',
    skillStage: 'development',
    skillType: SkillType.SMART_HOME,
  });
  skill.node.addDependency(endpointPermission);
  endpointPermission.configureSkillId(stack, 'SkillPermission', skill);

  expect(Template.fromStack(stack)).toMatchSnapshot();
});
