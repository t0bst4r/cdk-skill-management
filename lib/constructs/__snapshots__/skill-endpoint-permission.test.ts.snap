// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillEndpointPermission 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "17c16a3854838fd3ff4bda08146122a6701f33b9c86ae17f415ad0dc47a97544.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EndpointPermissionInitialSkillPermissionDF8BA4EA": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":lambda:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":function:MyEndpointFunction",
            ],
          ],
        },
        "Principal": "alexa-appkit.amazon.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SkillPermissionAddPermission916130FF": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "EndpointPermissionInitialSkillPermissionDF8BA4EA",
        "SkillPermissionAddPermissionCustomResourcePolicy40191D92",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"Lambda","action":"addPermission","parameters":{"FunctionName":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":lambda:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":function:MyEndpointFunction","StatementId":"",
              {
                "Fn::GetAtt": [
                  "EndpointPermissionInitialSkillPermissionDF8BA4EA",
                  "Id",
                ],
              },
              "","Principal":"alexa-connectedhome.amazon.com","Action":"lambda:InvokeFunction","EventSourceToken":"My-Skill-Id"},"physicalResourceId":{"id":"AddPermission-c8f28818e7e4b88e6f3f3de6c0841e1cf214663817-My-Skill-Id"}}",
            ],
          ],
        },
        "Delete": {
          "Fn::Join": [
            "",
            [
              "{"service":"Lambda","action":"removePermission","parameters":{"FunctionName":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":lambda:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":function:MyEndpointFunction","StatementId":"",
              {
                "Fn::GetAtt": [
                  "EndpointPermissionInitialSkillPermissionDF8BA4EA",
                  "Id",
                ],
              },
              ""},"physicalResourceId":{"id":"RemovePermission-c8f28818e7e4b88e6f3f3de6c0841e1cf214663817-My-Skill-Id"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "SkillPermissionAddPermissionCustomResourcePolicy40191D92": {
      "DependsOn": [
        "EndpointPermissionInitialSkillPermissionDF8BA4EA",
      ],
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "lambda:RemovePermission",
                "lambda:AddPermission",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":lambda:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":function:MyEndpointFunction",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SkillPermissionAddPermissionCustomResourcePolicy40191D92",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SkillPermissionRemovePermission83A11C0A": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "EndpointPermissionInitialSkillPermissionDF8BA4EA",
        "SkillPermissionRemovePermissionCustomResourcePolicyE0EF14B7",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"Lambda","action":"removePermission","parameters":{"FunctionName":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":lambda:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":function:MyEndpointFunction","StatementId":"",
              {
                "Fn::GetAtt": [
                  "EndpointPermissionInitialSkillPermissionDF8BA4EA",
                  "Id",
                ],
              },
              ""},"physicalResourceId":{"id":"RemovePermission-c8f28818e7e4b88e6f3f3de6c0841e1cf214663817-My-Skill-Id"}}",
            ],
          ],
        },
        "Delete": {
          "Fn::Join": [
            "",
            [
              "{"service":"Lambda","action":"addPermission","parameters":{"FunctionName":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":lambda:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":function:MyEndpointFunction","StatementId":"",
              {
                "Fn::GetAtt": [
                  "EndpointPermissionInitialSkillPermissionDF8BA4EA",
                  "Id",
                ],
              },
              "","Principal":"alexa-appkit.amazon.com","Action":"lambda:InvokeFunction"},"physicalResourceId":{"id":"AddPermission-c8f28818e7e4b88e6f3f3de6c0841e1cf214663817-no-skill"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "SkillPermissionRemovePermissionCustomResourcePolicyE0EF14B7": {
      "DependsOn": [
        "EndpointPermissionInitialSkillPermissionDF8BA4EA",
      ],
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "lambda:RemovePermission",
                "lambda:AddPermission",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":lambda:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":function:MyEndpointFunction",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SkillPermissionRemovePermissionCustomResourcePolicyE0EF14B7",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
